### ISSUE TITLE
Make sure Oceananigans is type stable

### ISSUE THREAD

**Issue by ali-ramadhan** [climate-machine/Oceananigans.jl#552]
To start testing Oceananigans at low-precision arithmetic and with alternative types with @milankl:

> I think the first thing I would need to do is making sure that Oceananigans is type stable. Meaning that you can pass a type T as an argument and this type is used throughout all computations. Even though you may only use Float64 atm, but having an entirely type-stable code that also means things like
> ```julia
> function f(A::AbstractMatrix)
>     ...
>     0.5*A[i,j]
>     ...
> end
> ```
> are written as
> ```julia
> function f(A::Array{T,2}) where T
>     ...
>     one_half = T(0.5)
>     one_half*A[i,j]
> ```
> , makes sure that we can use Sherlog64 to understand where the problems are.
>
> Ideally, all type conversion are explicitly written into the code rather than relying on promotion

**Comment by ali-ramadhan** [comment]
We use `FT(0.5)` and `zero(FT)` in a lot of places but never tested it outside of `Float32` and `Float64` so I'm sure there are places we missed.

For type-stable FFT we'll probably need something like https://github.com/milankl/coolFFT.jl

**Comment by milankl** [comment]
Hey everyone, as an example I can point you to [Juls.jl](https://github.com/milankl/Juls.jl) which can be executed as
```julia
julia> using Juls
julia> RunJuls(Float64);
Starting Juls on Thu, 05 Dec 2019 15:03:14 without output.
100% Integration done in 1.19s.

julia> RunJuls(Float32);
Starting Juls on Thu, 05 Dec 2019 15:03:19 without output.
100% Integration done in 0.72s.

julia> RunJuls(Float16);
Starting Juls on Thu, 05 Dec 2019 15:03:28 without output.
100% Integration done in 1min, 8s.
```
without relying on type promotion.

**Comment by milankl** [comment]
Concerning FFTs: 
I've started once a [conversation](https://github.com/JuliaMath/AbstractFFTs.jl/issues/34), but not sure how far we are now to have generic FFTs for any number format. We should check for what's happening with [FourierTransforms.jl](https://github.com/JuliaComputing/FourierTransforms.jl)

**Comment by ali-ramadhan** [comment]
Yeah that would be cool. Might also not be a bad idea to just use coolFFT.jl as it's <40 lines and seems to work (if we restrict ourselves to powers of 2 for `Nx`, `Ny`, `Nz`).

**Comment by milankl** [comment]
Problem with coolFFT.jl is that it's fairly slow as it doesn't allow for `plan_fft` and  `plan_rfft`, both should be supported by FourierTransforms.jl though, but I don't know how type-flexible they are.

**Comment by navidcy** [comment]
@ali-ramadhan did we resolve this?
